---
title: Title
author: "Anne Barranger & Sylvain Schmitts"
date: '`r Sys.Date()`'
output:
  bookdown::html_document2:
    number_sections: false
    toc: true
    toc_float: yes
  bookdown::word_document2: default
  bookdown::pdf_document2:
    includes:
      before_body: ./template/doc_prefix.tex
      in_header: ./template/preamble.tex
    number_sections: false
    toc: false
    keep_tex: true
linestretch: 1.5
# csl: /home/sylvain/Documents/Bibliography/csl/mee.csl
# bibliography: /home/sylvain/Documents/Bibliography/library.bib
link-citations: yes
---

```{r config, message=FALSE, warning=FALSE, include=FALSE}
rm(list = ls()); invisible(gc())
library(knitr)
library(kableExtra)
if(knitr:::is_html_output()) options(knitr.table.format = "html") 
if(knitr:::is_latex_output()) options(knitr.table.format = "latex") 
library(tidyverse)
library(broom)
library(bayesplot)
theme_set(bayesplot::theme_default())
opts_chunk$set( echo = F, message = F, warning = F, 
                fig.height = 8, fig.width = 12,
                cache = T, cache.lazy = F)
```

# Introduction

<!-- Studying biodiversity structure -->

Tropical forests host over half of worlwide biodiversity [@Scheffers2012]. Biodiversity can be divided into three fundamental levels: genetic, species and ecosystem. Understanding how is structured each level is a keypoint to apprehend its complexity and to potentially make forecasts for its evolution [@Messier2010a]. Interplay of individuals structuration within and among species, i.e. within ecosystem, enlightens processes of resources partitioning and the impacts of biotic and abiotic paramaters **Still not rellay clear**. To such an extent, functional traits studies spotlight different ecological strategies among and between species **Unclear**.

<!-- Definition of functional traits and ecological signification -->

Functional traits are defined as morpho-physio-phenological **(I'm not fan of the morpho-physio-pheno...)** traits impacting fitness indirectly through their effect on individual performance, which comprises growth, reproduction, and survival [@violle_let_2007]. Consequently **why?**, functional traits appear to be a perfect approach to study the effect of environment on the structure of intraspecific variability within population **what is population ?**.

<!-- Correlation of traits and signification -->

Many studies of functional traits highlighted general trends of variation between traits **Which variation ? Introduce here major known variation from litterature to be further discussed in discussion**. Those trade-offs between traits define different strategies to maximise individual performance. We'll focus on the way functional traits describe variation within diferent ecological scales, thus how they enable to distinguish thoses groups and what ecological consequences it might draw.

<!-- dealing with phenotype definition and its variation within ecological levels -->

**Introduction from A03-Models**

_Ecological community has been defined as a collection of species occurring at the same place, circumscribed by natural or arbitrary boundaries. Species within a community are presupposed to have intereactions, would they be genetical (ie gene flow) or ecological (ie competition). Therefore, these intereactions shape a variability in individuals morphological and physiological caracteristics **trop vague?**  within a community, that could lead to different ability of an ecological group to adapt in a specific environment. This variability could be assessed by a phenotype approach as phenotypic variation relies on genotypic information, ontogeny of individuals, environmental influences and random stochastic factors. Genetic heritage data wasn't available yet for these sets of analysis, but in a first approach, taxonomic levels within a community (ie species complex, species and individuals) are supposed to reflect this information. **taxonomic or ecological levels?** Phenotype description was assessed through a set of functionnal traits which are to be related with individuals abilities to grow, survive and reproduce in a given environment._

_We will develop a general form of model that will be able to best explain phenotype caracteristics (ie functionnal traits) according to environmental, ecological and ontogenitical inputs._

# Material & Methods

# Results
```{r data}
traitsSymphonia <- googlesheets::gs_title("Measures_Symphonia") %>%
  googlesheets::gs_read("AllTraits") %>%
  mutate(SLA = as.numeric(SLA), LDMC = as.numeric(LDMC)) %>%
  mutate(Genus = "Symphonia") %>%
  rename(Species = Morphotype) %>%
  mutate(Species = ifelse(Species == "Indet.",
                          c("globulifera", "sp.1", "sp.1")[fct_recode(Bark, "globulifera" = "G",
                                     "sp.1" =  "S")], Species))
traitsEschweilera <- googlesheets::gs_title("Measures_Eschweilera") %>%
  googlesheets::gs_read("AllTraits") %>%
  filter(!(Plot == 14 & SubPlot == 1 & TreeFieldNum == 760)) %>%  # outlier
  filter(!(Species %in% c("congestiflora","simiorum","persistens")))
traits <- bind_rows(traitsEschweilera, traitsSymphonia) %>%
  mutate(SpeciesLong = paste(substr(Genus, 1, 1), Species)) %>%
  group_by(idTree, Plot, SubPlot, TreeFieldNum, Genus, Species,
           SpeciesLong, Bark, Dawkins) %>%
  summarise_at(vars("SLA", "LDMC", "LT", "LA", "CC",
                    "brBT", "brBD", "brWD"), mean, na.rm = T) %>%
  ungroup() %>%
  mutate(invSLA=1/SLA) %>%
  mutate(invLA=1/LA)
load("./functional_save/env.Rdata")
Individuals <-  left_join(traits, env, by = "idTree", suffix = c("", ".y"))
rm(traitsEschweilera, traitsSymphonia, traits, paracou, env)
# save(Individuals, file = "./functional_save/Individuals.Rdata")
# load("./functional_save/Individuals.Rdata")
load("./functional_save/CompetitionMatrix.Rdata")
```

```{r mdata, fig.cap="prepared data",fig.width=12}
#traits <- c("LDMC")
#Individuals <- sample_n(Individuals, 100)
traits <- c("invSLA", "LDMC", "LT", "invLA", "CC")
mdata <- lapply(traits, function(trait){
  data_trait <- Individuals[!is.na(unlist(Individuals[,trait])),]
  list(N = nrow(data_trait),
       J = nrow(Competition),
       C = length(unique(data_trait$Genus)),
       S = length(unique(data_trait$Species)),
       Trait = as.numeric(scale(data_trait[trait], center = F)),
       DBH = as.numeric(scale(data_trait$DBH, center = F)),
       TWI = as.numeric(scale(data_trait$TWI, center = F)),
       individual = as.numeric(as.factor(Competition$idTree)),
       DBHj = as.numeric(scale(Competition$DBHj, center = F)),
       Deltaj = as.numeric(scale(Competition$dij, center = F)),
       species = as.numeric(as.factor(data_trait$Species)),
       complex = as.numeric(as.factor(data_trait$Genus)),
       speciesincomplex = unique(cbind(as.numeric(as.factor(data_trait$Species)), as.numeric(as.factor(data_trait$Genus))))[order(unique(cbind(as.numeric(as.factor(data_trait$Species)), as.numeric(as.factor(data_trait$Genus))))[,1]),2])})
names(mdata) <- traits

lapply(as.list(traits), function(trait)
  cbind(Model = trait, 
        Trait = mdata[[trait]]$Trait, 
        Individuals[unlist(!is.na(Individuals[trait])),])) %>% 
  bind_rows() %>% 
  select(Model, Trait, Genus, DBH, TWI) %>% 
  reshape2::melt(id.vars = c("Model", "Trait", "Genus")) %>% 
  ggplot(aes(x = value, y = Trait, col = Genus)) +
  geom_point() +
  geom_smooth() +
  facet_grid(Model ~ variable, scales = "free")
#rm(Individuals,Competition)
```

## Additive model on ontogeny and environment

### Model


\begin{equation} 
  T_{c,s,i} \sim \mathcal{N}(\alpha_{c,s}.\frac{DBH_i}{{\beta_{DBH}}_{c,s} + DBH_i} + {\beta_{TWI}}_{c,s}.TWI_i, \sigma)    \\
  \Theta_{c,s} \sim \mathcal{N}^5(\Theta_c,\sigma_{species}) ~|~ \Theta =
   \begin{bmatrix}
           \alpha \\
           \beta_{DBH} \\
           \beta_{TWI} \\
  \end{bmatrix},~\sigma_{species}
  \begin{bmatrix}
           \sigma_{Intercept} \\
           \sigma_{DBH} \\
           \sigma_{TWI} \\
  \end{bmatrix}  \\
  \Theta_c \sim \mathcal{N}^{3}(0, + \infty) \\
  (\sigma, \sigma_{species}) \sim \Gamma^{4}(- \infty, + \infty)
(\#eq:final)
\end{equation}

* $Trait_{c,s,i}$ is a trait of an individual i, from the species s, and the complex c;
* $i \in [1,I_s]$ where $I_s$ is the number of individuals in species s;
* $s \in [1,S_c]$ where $S_c$ is the number of species in complex c;
* $c \in [1,C]$ where $C$ is the number of complexes.
* $i \in [1,I]$ where I is the number of individuals;
* $j \in [1,J_i]$ where $J_i$ is the number of neighbors within the area of diameter $\delta_{max}$ surrounding the individual i, and $DBH_j$ the value of $DBH$ for neighbor $j$;
* $\alpha$ represents both the plateau of the Michaelis Menten and the intercept of each submodel (ontogeny, biotic & abiotic);
* $\beta_{DBH}$  is the value of DBH for which the trait account for half of its saturation value, and thus, ranges in positive values;
* $\beta_{TWI}$ is linear abiotic model parameters. $\beta_{TWI}$ can range in positive or negative value;
* $\beta_{Comp}$ is linear biotic model parameters. $\beta_{Comp}$ can range in positive or negative value;
* $\alpha_{NCI}$ represents the effect of neighbors distance to the individual and will be chosen positive;
* all $\sigma_{species}$ represents parameters variance for species;
* $\sigma$ and $\sigma_{species}$ are standard errors, and thus will have positive values;
* therefore, gamma laws were chosen as non informative flat priors on $\mathbb{R}^+$ for $\sigma$  and $\sigma_{species}$. Normal flat law were chosen as non informative flat prior on $\mathbb{R}$ for all $\Theta_{c,s}$ and $\Theta_c$.


### Summary

```{r Msummary}
load("./functional_save/AddOntoAbiot.Rdata")
pars <- c("alpha_c", "betaDBH_c", "betaTWI_c", "sigmaIntercept", "sigmaDBH", "sigmaTWI", "sigma")
lapply(as.list(traits), function(trait)
  broom::tidyMCMC(fits[[trait]], pars = c(pars, "lp__"), droppars = NULL, rhat = T) %>%
   mutate(Trait = trait)) %>%
  bind_rows() %>%
  kable() %>%
  kable_styling()
```

### Effects of complexes on parameters

```{r , fig.cap="Comparison of betaTWI.", eval=F, fig.height=8,fig.width=8}
plots <- lapply(traits,function(trait)
  {as.data.frame(fits[[trait]],pars =c("betaTWI_c[1]","betaTWI_c[2]")) %>% 
  reshape2::melt() %>% 
  ggplot(aes(x=variable,y=value,fill=variable)) +
  geom_boxplot() +
  scale_fill_discrete(name = "Complex", labels = c("Eschweilera","Symphonia")) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  labs(y = "DBH") +
  #coord_flip() +
  ggpubr::stat_compare_means(method = "t.test", aes(label = paste("T-test", ..p.signif..)))})

legend <- cowplot::get_legend(plots[[1]])

cowplot::plot_grid(plots[[1]] + theme(legend.position="none"),
                   plots[[2]] + theme(legend.position="none"),
                   plots[[3]] + theme(legend.position="none"),
                   plots[[4]] + theme(legend.position="none"),
                   plots[[5]] + theme(legend.position="none"),
                   legend,
                   labels = traits)
```


```{r}
include_graphics("./functional_save/AdditivePosteriors2.png")
```

Déréduire les données pour analyser de manière écologiques les paramètres.

### Analysis of variances

```{r Speciescode}
mdata$invSLA$species %>% 
  cbind(Individuals$SpeciesLong) %>% 
  unique() %>% 
  as_tibble() %>% 
  rename(Species=V1) %>% 
  kable() %>% 
  kable_styling()

# kable(as_tibble(unique(cbind(mdata$LDMC$species,Individuals$SpeciesLong))))
```

```{r SigmaIntercept}
cowplot::plot_grid(plotlist = lapply(traits, function(trait)
  mcmc_intervals(as.array(fits[[trait]]), regex_pars = "alpha_s")),
  labels = traits, nrow = 2)

mcmc_intervals(as.array(fits[["LDMC"]]), regex_pars = c("alpha_s"))
```

```{r sigma}
cowplot::plot_grid(plotlist = lapply(traits, function(trait)
  mcmc_areas(as.array(fits[[trait]]), regex_pars = "sigma")),
  labels = traits, nrow = 2)
```
```{r varianceanalysis, fig.cap="Table of coefficient of variation at the species level"}
means <- as.tibble(unlist(lapply(as.list(traits),function(trait)
  {mean(as.numeric(unlist(Individuals[trait])),na.rm=TRUE)
  })))
lapply(as.list(traits),function(trait){
  broom::tidyMCMC(fits[[trait]], pars = c(pars), droppars = NULL) %>%
    mutate(Trait = trait)}) %>% 
  bind_rows() %>%
  select(term,estimate,Trait) %>% 
  reshape2::dcast(Trait~term,value.var = "estimate") %>% 
  bind_cols(means) %>%
  mutate(CVsigma=sqrt(sigma)/abs(`value`)*100) %>%
  mutate(CVsigmaIntercept1=sqrt(sigmaIntercept)/abs(`alpha_c[1]`)*100) %>%
  mutate(CVsigmaIntercept2=sqrt(sigmaIntercept)/abs(`alpha_c[2]`)*100) %>%
  mutate(CVsigmaDBH1=sqrt(sigmaDBH)/abs(`betaDBH_c[1]`)*100) %>%
  mutate(CVsigmaDBH2=sqrt(sigmaDBH)/abs(`betaDBH_c[2]`)*100) %>%
  mutate(CVsigmaTWI1=sqrt(sigmaTWI)/abs(`betaTWI_c[1]`)*100) %>%
  mutate(CVsigmaTWI2=sqrt(sigmaTWI)/abs(`betaTWI_c[2]`)*100) %>%
  select(Trait,CVsigma,CVsigmaIntercept1,CVsigmaIntercept2,CVsigmaDBH1,CVsigmaDBH2,CVsigmaTWI1,CVsigmaTWI2) %>% 
  t() %>% 
  kable() %>% 
  kable_styling()
```

CV>100% : huge variation of species around parameters means

LDMC: 

* 
```{r}
include_graphics("./functional_save/AdditivePosteriors.png")
```

# References
