---
title: "Ecological genomics of niche exploitation and individual performance in tropical forest trees"
author: "Sylvain Schmitt\\textsuperscript{1}, Myriam Heuertz\\textsuperscript{2}, Bruno Hérault\\textsuperscript{3}, Niklas Tysklind\\textsuperscript{3}"
email: "sylvain.schmitt@agroparistech.fr"
institute: "\\textsuperscript{1}Bordeaux University, UMR BIOGECO \\textsuperscript{2}INRA, UMR BIOGECO \\textsuperscript{3}CIRAD-INRA, UMR ECOFOG"
longinstitute: "UMR BIOGECO - Bordeaux University/INRA"
web: "https://sylvainschmitt.netlify.com/"
logo: "images/ID.png"
QR: "images/QR.png"
backimg: "images/bkg70.png"
bibliofiles: "references.bib"
bibliography: "references.bib"
posteroptions: width=84.1,height=118.9,scale=1.25 #,grid # portrait
#posteroptions: width=110,height=90,scale=1.2 #,grid # landscape
headerheight: 17cm
# large, Large, LARGE, huge, Huge, veryHuge, VeryHuge, VERYHuge
titlefont: size=\veryHuge,series=\bfseries
authorfont: size=\huge
institutefont: size=\Large
knit: (function(input, encoding, make = TRUE) { source('tex/makefile-renderer.R', local = TRUE) })
---

%% smart
%% to=latex
%% template=tex/poster_night.tex
%% filter=tex/poster-filters.py
%% biblatex

```{r packages-and-options, cache=FALSE}
library(ggplot2)
library(knitr)
# library(tidyverse)
# library(raster)
knitr::opts_chunk$set(fig.width = 8, fig.height = 6)
```

[columns=2]

[column]

# Introduction

Biodiversity presents three nested levels, from **individuals**, over **species**, to **ecosystems**. **Intraspecific variability**, i.e. variability of individual genotypes and phenotypes, has been suggested to promote **local adaptation** of populations and to promote **species coexistence** at the community level, thus suggesting a role for intraspecific variability in the origin and maintenance of biodiversity. However, we still know little about the effects of intraspecific variability on population dynamics, species interactions and ecosystem processes. 

We propose to use tropical forest tree species complexes to explore the interaction of intraspecific genetic and phenotypic variability with the biotic and abiotic environment. This work represents the PhD research of Sylvain Schmitt, conducted in the long-term tropical forest monitoring plot and research station of Paracou, managed by the CIRAD in french Guiana.

# Conceptual framework

**Intraspecific variability** is expressed through phenotypic variation of the species composing communities. **Phenotypic variation** will be shaped (i) by **genetic heritage**, through genotypes, (ii) by the **environment** (both abiotic and biotic) with spatial and temporal heterogeneity, and (iii) by **random** stochastic factors (see figure **1**).

### Genetics

Genetic variability refers to the genetic differences we find within and among genomes of individuals within the population. Woody species have been shown to harbour more **genetic variation within populations** but less among populations than non woody species. **Local adaptations** are not random inside populations and result in a **structured genetic variability** within demes. The fields of **ecological genomics** and **association genetics** aim to relate genotypes to their habitat and phenotype, respectively.

### Environment

Trees interact with their environment mainly through resources acquisition and suffering predation. **Biotic interactions**, **hydrology**, **pedology** and **topography** are locally driving trees' access to resources. They are **spatially structured** and heterogeneous. Contrarily to hydrology, pedology and topography which are relatively stable over an individual's life time, biotic factors vary at the individual timescale. The **temporal component** of environment has been little exploited to our knowledge and might play an important role in population genetic and phenotypic variability, by challenging local adaptation.

### Phenotypes

**Local adaptation** reflects that local populations tend to have a **higher fitness in their native environment** than in other environments, resulting in increased **performance** of locally adapted individuals. Performance pertains to the ability to **survive**, **grow** and **reproduce** and can be directly measured (growth rate, mortality rate, monitoring of cohorts or allele frequencies) or approached with functional traits. Functional traits can be classified as relating to **biochemistry**, **physiology**, **anatomy**, **morphology**, or **phenology**.  Functional traits arise from expressed genes through transcripts. Consequently, **performance and functional traits** allow to assess **fitness**, and thus help to explore the link between **genetic variability** and **environmental heterogeneity** due to **local adaptation**.

```{r scheme, fig.cap="Local adaptation in a species complex living in sympatry. Different demes $G$ grow in sympatry in specific habitats $E$ along an environmental gradient. The interaction of local environment $E_i$ and genotype $G_i$ result in phenotype $P_i$. Phentoype $P_i$ feeds back to its local environment through biotic interactions. Temporal variation of the environement influences the phenotype of the established genotype.", fig.height=8, fig.width=16}
include_graphics("./images/PhD5.png")
```

# Acknowledgements

<!-- [columns=2] -->

<!-- [column] -->

We thanks the university of Bordeaux for a PhD grant to Sylvain Schmitt and acknowledge support of a grant from Investissement d'Avenir grants of the ANR (CEBA:ANR-10-LABX-25-01). We are grateful to Pascal Petronelli and the CIRAD inventory team for their work on tree inventories and botanical identification. Special thanks go to Saint-Omer and Josselin Cazal, Ilke Gelaldi, and Fabien Lehuede for their assistance during sampling in Paracou station. We also thank Émilie Ducouret for her work on functional traits.

<!-- [column] -->

<!-- ![](images/QR.png) -->

<!-- [/columns] -->

[column]

# Hypothesis

We want to use the present conceptual framework to test the following hypotheses:

**H1. Closely related species exploit locally different niches due to specific functional adaptations diverging from a common genetic basis.**

*Genome scans of species pairs growing in different environmental conditions (e.g. hydric) will be examined to reveal putative signs of differential adaptation to habitat in order to explore the phenotype-genotype association.*

**H2. Hybridization is an important factor for adaptive evolution.**

*Hybridization of closely related species results in new genetic complexes on which natural selection can act. We thus expect genome scans to reveal adaptive introgression. We expect (i) adaptive genes of primordial phenotypes to be more shared between closely related species than expected at random, (ii) performant individuals growing at the limit of their ecological niche to show higher levels of adaptive introgression.*

**H3. Individual performance, especially in tree growth, is shaped by genetic characteristics and (abiotic and biotic) environmental conditions, the latter of which might be potentially underevaluated for tropical tree species.**

*We will use several systems of closely related species to adjust and improve growth models of tropical trees, including individual phenotypic trait and genome scan data, to reveal mechanisms of adaptive evolution to niche and to estimate heritability of performance for tropical trees.*

**H4. Intraspecific genetic and phenotypic variability promotes species coexistence through local adaptation.**

*Merging results on niche exploitation (H1), hybridization (H2), and individual performance (H3) of several systems of closely related species, we want to explore the role of both intraspecific genetic and phenotypic variability on species coexistence within the community. We want to use several taxa of dominant species, and we expect local adaptation to play an important role in local species coexistence despite competition.*

# Symphonia first leads

*Symphonia globulifera* (Clusiaceae) represents a **species complex** in French Guiana, comprising two weakly differentiated morphotype living in differentiated habitats: *S. globulifera* (sensu stricto) preferentially grows in valley bottoms experiencing seasonal flooding and *S. sp.1* exploit a variety of drier habitats along well-drained slopes and hilltops. We explored in detail **habitat preference** of the two morphotypes in the Paracou research station, highlighting topographic wetness index as a good predictor of the distribution of morphotypes (see figure **2A**). We **sampled 402 adult trees** for genome scan and functional trait analysis. We are using transcriptomic data of a transplantation (Tysklind et al, in prep) to design a **sequence capture** experiment (see figure **2B**). Preliminary, functional trait analysis of leaf and wood traits suggests weak morphotype differentiation (see figure **2C**). Finally, we hope to integrate environmental, genetic, and phenotypic contribution to individual performance through an **individual growth model**  (see figure **2D**).

```{r environmentOverlap, fig.cap="Morphotypes distribution along the topographical wetness index (TWI)"}
path <- "~/Documents/BIOGECO/PhD/data/Paracou"
docpath <- "~/Documents/BIOGECO/PhD/documents/Symphonia/"
crs <- '+proj=longlat +datum=NAD83 +no_defs +ellps=GRS80 +towgs84=0,0,0' # global crs definition
load(file = file.path(path, 'trees', 'symphonia.Rdata'))
sympho <- sympho %>% 
  filter(campagne == 2015) %>% 
  filter(code_vivant == 1) %>% 
  filter(code_mesure == 0) %>% 
  filter(n_parcelle != 18) %>% 
  mutate(morphotype = espece)
symphoXY <- sympho
coordinates(symphoXY) <- ~Xutm + Yutm
proj4string(symphoXY) <- '+proj=utm +zone=22 +datum=WGS84 +units=m +no_defs +ellps=WGS84 +towgs84=0,0,0'
symphoXY <- spTransform(symphoXY, CRSobj = crs)
wetness <- raster(file.path(docpath, "./environment_save/env/wetness.tiff"))
dem <- raster(file.path(path, "topography", "MNT_ParacouAvril2009_5m.tif")) # for CRS
projection(wetness) <- projection(dem)
wetness <- projectRaster(wetness, crs = crs)
sympho$wetness <- raster::extract(wetness, symphoXY)
ge <- sympho %>% 
  ggplot(aes(wetness, col = morphotype, fill = morphotype)) +
  geom_histogram(aes(y = ..density..), binwidth = 0.05,
                 alpha = 0.1, position = "dodge") +
  geom_density(alpha = 0.2) +
  xlab('Tropographic wetness index') +
  bayesplot::theme_default()
```

```{r transciptomic, fig.cap="Candidate SNPs for sequence capture from Tysklind et al, in prep."}
path <- "~/Documents/BIOGECO/PhD/data/Symphonia_Niklas/k2rt"
load(file.path(path, "count_SNP.Rdatat"))
gg <- count_SNP %>% 
  ungroup() %>% 
  mutate(type = ifelse(Is_in_CDS == "False" & Is_not_synonymous == "N/A", 
                       "SNP in untranslated\nregion (UTR)", NA)) %>% 
  mutate(type = ifelse(Is_in_CDS == "True" & Is_not_synonymous == "False", 
                       "Synonymous SNP in\ncoding region", type)) %>% 
  mutate(type = ifelse(Is_in_CDS == "True" & Is_not_synonymous == "True", 
                       "SNP in coding region\nwith functional impact", type)) %>% 
  filter(!is.na(type)) %>% 
  ggplot(aes(type, n, fill = Is_condition_specific)) +
  geom_bar(stat = "identity", position = "dodge") +
  coord_flip() +
  scale_y_continuous(labels = function(x) format(x, big.mark = " ")) +
  xlab(" ") + scale_x_discrete(limits = c("SNP in coding region\nwith functional impact",
                                          "Synonymous SNP in\ncoding region",
                                          "SNP in untranslated\nregion (UTR)")) +
  scale_fill_discrete("Morphotype-specific", labels = c("no", "yes")) +
  geom_text(aes(y = n + 10^5, 
                col = Is_condition_specific,
                label = paste(format(n, big.mark = " "))), 
            size = 3,
            position = position_dodge(width = 1)) +
  scale_color_discrete(guide = "none") +
  bayesplot::theme_default()
```

```{r functionalOverlap, fig.cap="Morphotypes distribution in SLA-BrWD scatterplot."}
path <- "~/Documents/BIOGECO/PhD/data/Symphonia_Paracou/"
load(file.path(path, "data.Rdata"))
gf <- data  %>% 
  dplyr::select(morphotype, Dry_Weight, Area_exclude, Wood_infra_density) %>% 
  mutate(SLA = Area_exclude / Dry_Weight) %>% 
  mutate(WD = Wood_infra_density) %>% 
  na.omit() %>% 
  ggplot(aes(WD, SLA, col = morphotype)) +
  geom_point() +
  stat_ellipse() +
  xlab("Branch wood density (BrWD in g/g)") +
  ylab("Specific leaf area (SLA in cm/g)") +
  bayesplot::theme_default()
```

```{r individualGrowth, fig.cap="Symphonia growth model with individual maximum growth rate (Gmax)."}
docpath <- "~/Documents/BIOGECO/PhD/documents/Symphonia/growth/"
mpath <- "./growth_models"
load(file.path(docpath, "./growth_save/model_data.Rdata"))
load(file.path(docpath, mpath, 'Nio2.Rdata'))
pars <- c("AGRmax", "Dopt", "Ks", "sigma", "sigma_ind")
gp <- data.frame(
  mean_all = apply(as.matrix(fit, pars = "AGRpred"), 2, mean),
  mean_ind = apply(as.matrix(fit, pars = "AGRpred_ind"), 2, mean),
  t(apply(as.matrix(fit, pars = "AGRpred_ind"), 2, quantile, probs = c(0.05, 0.95))),
  data2$trees
) %>% 
  filter(agr < 2) %>% 
  ggplot(aes(dbh_c, agr)) +
  geom_point(aes(col = espece)) +
  geom_line(aes(y = mean_ind, group = idArbre), alpha = 0.4) +
  geom_ribbon(aes(ymin = X5., ymax = X95., group = idArbre), color = 'grey', alpha = 0.05) +
  geom_line(aes(y = mean_all), lwd = 1.2) +
  ylab("Annual growth rate (AGR in cm/yr)") +
  xlab("Diameter at breast height (dbh in cm)") +
  scale_color_discrete("morphotype") +
  bayesplot::theme_default()
```

```{r allSympho, fig.cap="Preliminary results for the \\emph{Symphonia globulifera} study case. Subplot \\textbf{A} shows the distribution of morphotypes along the topographical wetness index. Subplot \\textbf{B} shows candidate SNPs for sequence capture from Tysklind et al, in prep. Subplot \\textbf{C} shows the distibution of  according to specific leaf area and branch wood density. Subplot \\textbf{D} shows a growth model with individual maximum growth rate.", fig.height=12, fig.width=16}
cowplot::plot_grid(ge, gg, gf, gp, labels = LETTERS[1:4]) 
```

<!-- ![](images/QR.png) -->
<!-- \raggedleft\includegraphics[height=\headerheight]{$QR$} -->

[/columns]
