```{r setup_intro, include=FALSE}
rm(list = ls()) ; invisible(gc()) ; set.seed(42)
library(tidyverse)
```

# Introduction {-}

This document review and summarize sequence capture and genomic analysis of *Symphonia globulifera*.

## Modus operandi (to discuss):

We will proceed sequence capture and genomic analysis of *Symphonia globulifera* with following steps:

- [ ] Choose kit (MyBaits ?, NEBnext ?)
    - [ ] Total sequence length ?
    - [ ] Number of sequences ?
    - [ ] Sequences depth (O.5-1 kbp ?) ?
- [ ] Select sequence to be captured (see figure \@ref(fig:seqTree))
    - [ ] Morphtoype specific genes (Tysklind et al)
    - [ ] Non specific genes (Tysklind et al)
    - [ ] Intron & UTR (Tysklind et al, Scotti et al, or African *S. globulifera*)
    - [ ] Chloroplast sequences (Scotti et al)
- [ ] Design probes (software with the kit)
- [ ] Extract DNA
    - [x] Prepare leaf extracts
    - [ ] Review, update and reprint protocol
- [ ] Wait libraries
- [ ] Raw data analysis
    - [ ] Genome by Sequencing (GbS)
    - [ ] Reads mapping
    - [ ] SNP call
    - [ ] SNP filtering
- [ ] Genetic data analysis
    - [ ] Genetic structure of populations
    - [ ] Genomic structure of polymorphisms
    - [ ] Ecological genomic (genotype-environment association)
    - [ ] Assocaition genomic (genotype-phenotype association)

```{r seqTree, fig.cap="Different types in which to choose sequences to capture.", echo=F}
DiagrammeR::DiagrammeR("
   graph TB
   DNA-->Nucleus
   DNA-->Chlorplast
   Nucleus-->Neutral
   Neutral-->Intron
   Neutral-->UTR
   UTR-->Specific
   UTR-->Non_Specific
   Nucleus-->CDS
   CDS-->Synonymous
   CDS-->Non_Synonymous
   Synonymous-->Specific
   Synonymous-->Non_Specific
   Non_Synonymous-->Specific
   Non_Synonymous-->Non_Specific
")
   # Chloroplast-->Scoti_et_al
   # Intron-->Scoti_et_al
   # Specific-->Niklas_et_al
   # Non_Specific-->Niklas_et_al
```

## Sequence selection

We have to choose sequences to be captured. The sequence selection for capture will be detailed in next chapters. We will use 5 sources of information:

* __Transcriptomic analysis from Tysklind *et al (in prep)*__, realized on transplant experiment of *Symphonia globulifera* in French Guiana this study will inform us on untrasnlated regions (UTR) and coding DNA sequences (CDS) precizing the putative functional impact and the morphotype specificity of genes
* __SNPs from French Guiana obtain by GbS in @Torroba-Balmori2017 __, will inform us on guianan SNPs for neutral regions
* __genome scaffolds from Africa in @Olsson2017 __, will bring us neutral scaffold
* __chloroplast sequences from French Guiana on NCBI__, will brings us short sequences of chloroplast
* __chlorplast assembly from Ameria from Bautista et Claros__, will bring us genomic sequences

We will select approxamitely:

* __60%__ of coding sequence with a functional impact with data from Tysklind *et al (in prep)* split between:
    * __40%__ of sequences including morphotype-specific variants
    * __20%__ of sequences including non morphotype-specific variants
* __30%__ of genomic neutral regions (introns and UTR with some including morphotype-specific variants) from both data from Tysklind *et al (in prep)* and a interpolation between genomic data from @Torroba-Balmori2017 and @Olsson2017
* __10%__ of chloroplastic regions from interpolation between Bautista et Claros and NCBI american sequences
    
```{r planCapture, fig.cap = "Sequences to be captured.", echo=FALSE}
data.frame(
  category = c("CDS-MtSpecific", "CDS-NonSpecific", "Neutral-Neutral", "Chloroplast-Chloroplast"),
  size = c(40, 20, 30, 10)
) %>% 
  sunburstR::sunburst()
```