```{r setup_intro, include=FALSE}
rm(list = ls()) ; invisible(gc()) ; set.seed(42)
```

# Introduction {-}

## Modus operandi (to discuss):

- [ ] Choose kit (MyBaits ?, NEBnext ?)
    - [ ] Total sequence length ?
    - [ ] Number of sequences ?
    - [ ] Sequences depth (O.5-1 kbp ?) ?
- [ ] Select sequence to be captured (see figure \@ref(fig:seqTree))
    - [ ] Morphtoype specific genes (Tysklind et al)
    - [ ] Non specific genes (Tysklind et al)
    - [ ] Intron & UTR (Tysklind et al, Scotti et al, or African *S. globulifera*)
    - [ ] Mitochondrial sequences (Scotti et al)
- [ ] Design probes (software with the kit)
- [ ] Extract DNA
    - [ ] Prepare leaf extracts
    - [ ] Review, update and reprint protocol
- [ ] Wait libraries
- [ ] Raw data analysis
    - [ ] Genome by Sequencing (GbS)
    - [ ] Reads mapping
    - [ ] SNP call
    - [ ] SNP filtering
- [ ] Genetic data analysis
    - [ ] Genetic structure of populations
    - [ ] Genomic structure of polymorphisms
    - [ ] Ecological genomic (genotype-environment association)
    - [ ] Assocaition genomic (genotype-phenotype association)
    
```{r seqTree, fig.cap="Different types in which to choose sequences to capture.", echo=F}
DiagrammeR::DiagrammeR("
   graph TB
   DNA-->Nucleus
   DNA-->Chlorplast
   Nucleus-->Neutral
   Neutral-->Intron
   Neutral-->UTR
   UTR-->Specific
   UTR-->Non_Specific
   Nucleus-->CDS
   CDS-->Synonymous
   CDS-->Non_Synonymous
   Synonymous-->Specific
   Synonymous-->Non_Specific
   Non_Synonymous-->Specific
   Non_Synonymous-->Non_Specific
")
   # Chloroplast-->Scoti_et_al
   # Intron-->Scoti_et_al
   # Specific-->Niklas_et_al
   # Non_Specific-->Niklas_et_al
```

## Polymorphisms

What we have:

* if the variant is in a CDS (Coding DNA Sequence)
* if the variant is non-synonymous (Putative functional impact)
* if the variant is morphtoype-specific (Effectively at juvenile stage)
* Annotation

So what we need/can have is (if I try to summarize):

* SNP in CDS
    * Non-synonymous SNP
        * Morphotype-specific
        * Non specific
    * Synonymous SNP
        * Morphotype-specific
        * Non specific
* SNP in Untranslated Region (UTR)
    * Morphotype-specific
    * Non specific
* SNP in introns (cf Scoti et al)
* SNP in plastid DNA (cf ?)
