```{r setup_transcripto, include=FALSE}
rm(list = ls()) ; invisible(gc()) ; set.seed(42)
library(knitr)
library(tidyverse)
opts_chunk$set(
  echo = F, message = F, warning = F, fig.height = 6, fig.width = 8,
  cache = T, cache.lazy = F)
theme_set(bayesplot::theme_default())
path <- "~/Documents/BIOGECO/PhD/data/Symphonia_Niklas/k2rt"
```

# Transcriptomic data

Niklas used 20 Symphonia juveniles from the transplantation garden experiment for transcriptomic analysis. RNA sequence were captured. The analysis followed the scheme suggested by @Lopez-Maestre2016 (see below). First, reads were assembled with `Trinity` into transcripts. In parrallel, SNPs were detected with `Kissplice`. Then SNPs have been mapped on the transcritpome with `BLAT`. In parrallel SNPs have been tested to be morphotype-specific at the level $\alpha = 0.001$ with `KissDE` and transcriptome Open Reading Frames (ORF) have been indentified with `Transdecoder`. Finally, SNPs functional impact have been evaluated through `k2rt`. Consequently, for every SNP we have the following informations: (i) inside coding DNA sequence (CDS), (ii) synonymous or not, (iii) morphotype-specificity.

![Analysis scheme from @Lopez-Maestre2016.](./probes_data/analysisScheme.jpeg)

## Quality check

We assessed transcriptomic analysis quality with possible sequencing errors, and SNPs in multiple assembled genes or isoforms (see table \@ref(tab:Qcheck)). We find...

```{r Qcheck}
data <- read_delim(file.path(path, "symphonia_juv_fullsample_trinity500_k2rt_type_0a_mainOutput.tsv"),
           delim = "\t")
data %>% 
  select(Possible_sequencing_error,
         SNP_in_mutliple_assembled_genes,
         SNP_in_mutliple_assembled_isoforms) %>% 
  mutate(Count = 1) %>%
  reshape2::melt(id.vars = "Count") %>% 
  filter(value == "True") %>% 
  mutate(variable = gsub("_", " ", variable)) %>% 
  group_by(variable) %>% 
  summarise(n = n()) %>% 
  mutate(Percentage = round(n/nrow(data)*100,2)) %>% 
  kable(caption = "Quality check with single SNPs....",
        format.args = list(big.mark = " "))
rm(data) ; invisible(gc())
```

## SNPs

Once quality check, we can look at SNP distribution between the three modalities (see table \@ref(tab:countTable) and and \@ref(fig:sunburstcount)). We can see... More particularly we have 99 491 Non-synonymous SNPs in coding region that are morphotype specific making a pool of SNPs to be used in sequence capture. We now need to identify gene functions for those SNPs (GO terms) and more particularly if some functions are enriched in a particular morphotype.

```{r count_SNP}
data <- read_delim(file.path(path, "symphonia_juv_fullsample_trinity500_k2rt_type_0a_mainOutput.tsv"),
           delim = "\t")
count_SNP <- data %>% 
  filter(Possible_sequencing_error != "True") %>% 
  select(Is_in_CDS, Is_not_synonymous, Is_condition_specific) %>% 
  group_by(Is_in_CDS, Is_not_synonymous, Is_condition_specific) %>% 
  summarize(n = n()) 
rm(data) ; invisible(gc())
```

```{r countTable}
count_SNP %>% 
  dplyr::rename("Coding sequence" = Is_in_CDS,
         "Not synonymous" = Is_not_synonymous,
         "Morphotype-specific" = Is_condition_specific) %>% 
  kable(caption = "Single SNPs headcount for Symphonia juveniles in k2rt main output. First column indicates if the SNP is in a coding sequence, second column indicates is the SNP is not synonymous, third column indicates if the SNP is morphotype-specific, and fourth column indicates the headcount.", format.args = list(big.mark = " "))
```

<!-- # ```{r countGraph, fig.cap="SNPs headcount for Symphonia juveniles by types. Bar fill color indicates if the SNP is morphotype specific (blue if yes, red if not)."} -->
<!-- # count_SNP %>%  -->
<!-- #   ungroup() %>%  -->
<!-- #   mutate(type = ifelse(Is_in_CDS == "False" & Is_not_synonymous == "N/A",  -->
<!-- #                        "SNP in untranslated\nregion (UTR)", NA)) %>%  -->
<!-- #   mutate(type = ifelse(Is_in_CDS == "True" & Is_not_synonymous == "False",  -->
<!-- #                        "Synonymous SNP in\ncoding region", type)) %>%  -->
<!-- #   mutate(type = ifelse(Is_in_CDS == "True" & Is_not_synonymous == "True",  -->
<!-- #                        "SNP in coding region\nwith functional impact", type)) %>%  -->
<!-- #   filter(!is.na(type)) %>%  -->
<!-- #   ggplot(aes(type, n, fill = Is_condition_specific)) + -->
<!-- #   geom_bar(stat = "identity", position = "dodge") + -->
<!-- #   coord_flip() + -->
<!-- #   scale_y_continuous(labels = function(x) format(x, big.mark = " ")) + -->
<!-- #   xlab(" ") + scale_x_discrete(limits = c("Not synonymous SNP\nin coding region", -->
<!-- #                                           "Synonymous SNP in\ncoding region", -->
<!-- #                                           "SNP in untranslated\nregion (UTR)")) + -->
<!-- #   scale_fill_discrete("Morphotype-specific", labels = c("no", "yes")) + -->
<!-- #   geom_text(aes(y = n + 10^5,  -->
<!-- #                 col = Is_condition_specific, -->
<!-- #                 label = paste(format(n, big.mark = " "))),  -->
<!-- #             size = 3, -->
<!-- #             position = position_dodge(width = 1)) + -->
<!-- #   scale_color_discrete(guide = "none") -->
<!-- # ``` -->

```{r sunburstcount, fig.cap = "SNPs headcount for Symphonia juveniles by types. Levels precise if the SNP is in a coding region, synonymous and finally morphotype-specific."}
count_SNP %>% 
  ungroup() %>%
  mutate(type = ifelse(Is_in_CDS == "False" & Is_not_synonymous == "N/A", 
                       "UTR-UTR-", NA)) %>% 
  mutate(type = ifelse(Is_in_CDS == "True" & Is_not_synonymous == "False", 
                       "CDS-synonymous-", type)) %>% 
  mutate(type = ifelse(Is_in_CDS == "True" & Is_not_synonymous == "True", 
                       "CDS-not synonymous-", type)) %>% 
  filter(!is.na(type)) %>% 
  mutate(Is_condition_specific = ifelse(Is_condition_specific == "False",
                                        "non specific", "morphotype specific")) %>% 
  mutate(type = paste0(type,Is_condition_specific)) %>% 
  select(type, n) %>% 
  sunburstR::sunburst(count = T)
rm(list = ls())
```

## Genes

```{r total_genes}
data <- read_delim(file.path(path, "symphonia_juv_fullsample_trinity500_k2rt_type_0a_mainOutput.tsv"),
           delim = "\t")

total_genes <- data %>% 
  filter(Possible_sequencing_error != "True") %>% 
  dplyr::rename(gene =  `#Component_ID`) %>% 
  group_by(gene) %>% 
  summarize(n = n()) %>% 
  arrange(desc(n))

rm(data) ; invisible(gc())
```

We have a total of __`r format(nrow(total_genes), big.mark = " ")`__ genes with variable SNP count (from 398 to 1, see figure \@ref(fig:totalGraph)). We can see that almost all  so-called "genes" have a coding DNA sequence (CDS, see figure \@ref(fig:geneGraph)). Consequently we can focus on those genes that will contain both neutral and functional regions. Only half of genes, c.a. 120 k, have non-synonymous SNPs. More specifically, as a first choice we can focus on genes contianing SNPs in every categories, i.e. both morphotype-specific or not and synonymous or not, representing 42 966 genes (see figure \@ref(fig:genesVen)).

```{r totalGraph, fig.cap="SNP rank-abundance by gene."}
total_genes %>%
  ggplot(aes(1:nrow(total_genes), n)) +
  geom_linerange(ymin = 0, aes(ymax = n)) +
  ylab("Number of SNPs") +
  xlab("gene")
```


```{r genesCount}
data <- read_delim(file.path(path, "symphonia_juv_fullsample_trinity500_k2rt_type_0a_mainOutput.tsv"),
           delim = "\t")

countGenes <- data %>% 
  filter(Possible_sequencing_error != "True") %>% 
  dplyr::rename(gene =  `#Component_ID`,
         SNP = SNP_ID,
         CDS = Is_in_CDS,
         NotSyn = Is_not_synonymous,
         MtSpec = Is_condition_specific) %>% 
  select(gene, CDS, NotSyn, MtSpec) %>% 
  unique() %>% 
  group_by(CDS, NotSyn, MtSpec) %>% 
  summarize(n = n()) %>% 
  ungroup() 

rm(data) ; invisible(gc())
```

```{r, geneGraph, fig.cap = "gene headcount for Symphonia juveniles by types. Levels precise if the gene contain SNP in a coding region, synonymous and finally morphotype-specific."}
countGenes %>% 
  mutate(type = ifelse(CDS == "False" & NotSyn == "N/A", 
                       "UTR-UTR-", NA)) %>% 
  mutate(type = ifelse(CDS == "True" & NotSyn == "False", 
                       "CDS-synonymous-", type)) %>% 
  mutate(type = ifelse(CDS == "True" & NotSyn == "True", 
                       "CDS-not synonymous-", type)) %>% 
  filter(!is.na(type)) %>% 
  mutate(MtSpec = ifelse(MtSpec == "False",
                         "non specific", "morphotype specific")) %>% 
  mutate(type = paste0(type,MtSpec)) %>% 
  select(type, n) %>% 
  sunburstR::sunburst(count = T)
```

```{r genesVen, fig.cap="Genes distribution between depending on the type of their SNPs. Gene including SNP with coding DNA sequences (CDS), functional synonymy, or morphtype pspecificity."}
data <- read_delim(file.path(path, "symphonia_juv_fullsample_trinity500_k2rt_type_0a_mainOutput.tsv"),
           delim = "\t") %>% 
  filter(Possible_sequencing_error != "True")
venn.genes <- list(Synonymous = data$`#Component_ID`[data$Is_not_synonymous == "False" &
                                                    data$Is_in_CDS == "True"],
                  "Not Synonymous" = data$`#Component_ID`[
                    data$Is_not_synonymous == "True" &
                      data$Is_in_CDS == "True"],
                  "Morphotype-specific" = data$`#Component_ID`[
                    data$Is_condition_specific == "True" & data$Is_in_CDS == "True"],
                  "Not Morphotype-specific" = data$`#Component_ID`[
                    data$Is_condition_specific == "False" & data$Is_in_CDS == "True"])
venn.graph <- VennDiagram::venn.diagram(venn.genes , NULL,
                                        fill = RColorBrewer::brewer.pal(4, "Accent"),
                                        alpha = 0.5)
grid.draw(venn.graph)
```

We will use 

## Gene ontology

<!-- # ```{r GOenriched} -->
<!-- # OE <- sapply(list.files(path_GO),  -->
<!-- #        function(file) read_tsv(file.path(path_GO, file)), -->
<!-- #        simplify = F) %>%  -->
<!-- #   bind_rows(.id = "id") -->
<!-- # ``` -->
