```{r setup_genomic_results, include=FALSE}
rm(list = ls()) ; invisible(gc()) ; set.seed(42)
library(knitr)
library(parallel)
library(Biostrings)
library(tidyverse)
theme_set(bayesplot::theme_default())
opts_chunk$set(
  echo = F, message = F, warning = F, fig.height = 6, fig.width = 8,
  cache = T, cache.lazy = F)
path <- "~/Documents/BIOGECO/PhD/data/Symphonia_Genomes/"
```

# Genomic data - Results

## Scaffolds from Scotti *et al (in prep)*

### Consensus sequences blast

In total **542 scaffolds** from Scotti *et al (in prep)* match consensus sequences from Torroba-Balmori et al. (unpublished). But several scaffolds obtained multiple matches that we cannot use for probes. We will thus exclude the whole scaffold if the scaffold is shorter than 2000 bp or the scaffold region matching the raw read if the scaffold is longer than 2000 bp.

```{r blastResultScottiFig, fig.cap="Number of match with Torroba consensus reads vs gene width.", fig.height=4}
blast_Ivan <- read_tsv(file.path(path, "Ivan_2018", "torroba_blast", "blast_consensus_torroba2.txt"), col_names = F)
names(blast_Ivan) <- c("Read", "Scaffold", "Perc_Ident", "Alignment_length", "Mismatches",
                  "Gap_openings", "R_start", "R_end", "S_start", "S_end", "E", "Bits")
scf_Ivan <- readDNAStringSet(file.path(path, "Ivan_2018", "torroba_blast", "selected_scaffolds_blast_consensus2.fa"))
blast_Ivan %>%
  dplyr::left_join(data.frame(Scaffold = names(scf_Ivan), width = width(scf_Ivan))) %>% 
  select(Scaffold, Read, width) %>%
  unique() %>%
  group_by(Scaffold, width) %>%
  summarise(n = n()) %>% 
  ggplot(aes(width, n)) +
  geom_point() +
  geom_jitter() +
  ylab("Number of match with consensus sequences from Torroba-Balmori et al (unpublished)") +
  xlab("Width of scaffolds from Scotti et al. (in prep) (in bp)")
```

```{r blastResultScottiFilter}
sel_Ivan <- blast_Ivan %>%
  dplyr::left_join(data.frame(Scaffold = names(scf_Ivan), width = width(scf_Ivan))) %>% 
  select(Scaffold, Read, width, S_start, S_end) %>% 
  group_by(Read) %>%
  filter(n() > 2) %>% 
  ungroup() %>% 
  mutate(remove = ifelse(width < 2000, "X", " ")) %>% 
  mutate(cut = ifelse(width > 2000, paste0(S_start, "-", S_end), " ")) %>% 
  select(Scaffold, width, remove, cut)
sel_Ivan <- list(rm = gsub("Ivan_2018_sympho47_", "", sel_Ivan$Scaffold[sel_Ivan$remove == "X"]),
                 cut = sel_Ivan[sel_Ivan$remove == " ",])
kable(sel_Ivan$cut, caption = "Scaffold to cut due to multiple read match.")
```

Following scaffolds will be removed due to multitple matches and a length $<200bp$: `r paste0(sel_Ivan$rm, collapse = ", ")`. And other will be cut (see table \@ref(tab:blastResultScottiFilter)).

### Raw reads alignment

## Scaffolds from @Olsson2017

### Consensus sequences blast

We finally have most of scaffolds with on match in a broad range of sizes (from 100 bp to 33.2 kbp). In total **688** scaffolds from @Olsson2017 match consensus sequences from Torroba-Balmori *et al. (unpublished)*. But several scaffolds obtained multiple matches that we cannot use for probes. We will thus exclude the whole scaffold if the scaffold is shorter than 2000 bp or the scaffold region matching the raw read if the scaffold is longer than 2000 bp.

```{r blastResultOlssonFig, fig.cap="Number of match with Torroba consensus reads vs gene width.", fig.height=4}
blast_Olsson <- read_tsv(file.path(path, "Olsson_2016", "torroba_blast", "blast_consensus_torroba2.txt"), col_names = F)
names(blast_Olsson) <- c("Read", "Scaffold", "Perc_Ident", "Alignment_length", "Mismatches",
                  "Gap_openings", "R_start", "R_end", "S_start", "S_end", "E", "Bits")
scf_Olsson <- readDNAStringSet(file.path(path, "Olsson_2016", "torroba_blast", "selected_scaffolds_blast_consensus2.fa"))
blast_Olsson %>%
  dplyr::left_join(data.frame(Scaffold = names(scf_Olsson), width = width(scf_Olsson))) %>% 
  select(Scaffold, Read, width) %>%
  unique() %>%
  group_by(Scaffold, width) %>%
  summarise(n = n()) %>% 
  ggplot(aes(width, n)) +
  geom_point() +
  geom_jitter() +
  ylab("Number of match with consensus sequences from Torroba-Balmori et al (unpublished)") +
  xlab("Width of scaffolds from 0lsson et al. (2017) (in bp)")
```

```{r blastResultOlssonFilter}
sel_Olsson <- blast_Olsson %>%
  dplyr::left_join(data.frame(Scaffold = names(scf_Olsson), width = width(scf_Olsson))) %>% 
  select(Scaffold, Read, width, S_start, S_end) %>% 
  group_by(Read) %>%
  filter(n() > 2) %>% 
  ungroup() %>% 
  mutate(remove = ifelse(width < 2000, "X", " ")) %>% 
  mutate(cut = ifelse(width > 2000, paste0(S_start, "-", S_end), " ")) %>% 
  select(Scaffold, width, remove, cut)
sel_Olsson <- list(rm = gsub("Olsson_2017_", "", sel_Olsson$Scaffold[sel_Olsson$remove == "X"]),
                 cut = sel_Olsson[sel_Olsson$remove == " ",])
kable(sel_Olsson$cut, caption = "Scaffold to cut due to multiple read match.") 
```

Following scaffolds will be removed due to multitple matches and a length $<200bp$: `r paste0(sel_Olsson$rm, collapse = ", ")`. And other will be cut (see table \@ref(tab:blastResultOlssonFilter)).

### Raw reads alignment

### Scotti *et al (in prep)* overlap

## Transcriptome alignment

### On Scotti *et al (in prep)*

We obtained **169 768** transcripts Tysklind *et al (in prep)* representing **60 184** candidate genes matching on **49 330** scaffolds from Scotti *et al (in prep)* (see fig \@ref(fig:bwaTranscriptResultScottiFig)). But the majority of transcripts have multiple matches on several scaffolds.

```{r bwaTranscriptResultScottiFig, fig.cap="Number of match with Torroba consensus reads vs gene width.", fig.height=4}
bed_Ivan <-  read_tsv(file.path(path, "Ivan_2018", "transcript_alignment", "filtered_transcripts.bed"), col_names = F) %>% 
  unique()
names(bed_Ivan) <- c("scaffold",  "start",  "end",    "transcript",   "score",  "strand") 
scf_Ivan <- readDNAStringSet(file.path(path, "Ivan_2018", "transcript_alignment", "selected_scaffolds.fa"))
trsc_Ivan <- readDNAStringSet(file.path(path, "Ivan_2018", "transcript_alignment", "selected_transcripts.fa"))
data_Ivan <- bed_Ivan %>%
  dplyr::left_join(data.frame(transcript = names(trsc_Ivan), width_trsc = width(trsc_Ivan))) %>% 
  dplyr::left_join(data.frame(scaffold = names(scf_Ivan), width_scf = width(scf_Ivan))) %>% 
  group_by(transcript) %>% 
  mutate(n_match = n()) %>% 
  ungroup()
g1 <- ggplot(data_Ivan, aes(width_scf, width_trsc, col = n_match)) +
  geom_point() +
  xlab("Width of scaffolds from Scotti et al. (in prep) (in bp)") +
  ylab("Width of transcripts from Tysklind et al. (in prep) (in bp)") +
  scale_color_continuous("Matches")
g2 <- ggplot(data_Ivan, aes(n_match)) + 
  geom_histogram(binwidth = 1) +
  xlab("Matches") +
  coord_flip()
cowplot::plot_grid(g1, g2, nrow = 1, rel_widths = c(3,1))
```

### On @Olsson2017

We obtained **174 454** transcripts Tysklind *et al (in prep)* representing **61 290** candidate genes matching on **132 394** scaffolds from Scotti *et al (in prep)* (see fig \@ref(fig:bwaTranscriptResultOFigsson)). But the majority of transcripts have multiple matches on several scaffolds.

```{r bwaTranscriptResultOlssonFig, fig.cap="Number of match with Torroba consensus reads vs gene width.", fig.height=4}
bed_Olsson <-  read_tsv(file.path(path, "Olsson_2016", "transcript_alignment", "filtered_transcripts.bed"), col_names = F) %>% 
  unique()
names(bed_Olsson) <- c("scaffold",  "start",  "end",    "transcript",   "score",  "strand") 
scf_Olsson <- readDNAStringSet(file.path(path, "Olsson_2016", "transcript_alignment", "selected_scaffolds.fa"))
trsc_Olsson <- readDNAStringSet(file.path(path, "Olsson_2016", "transcript_alignment", "selected_transcripts.fa"))
data_Olsson <- bed_Olsson %>%
  dplyr::left_join(data.frame(transcript = names(trsc_Olsson), width_trsc = width(trsc_Olsson))) %>% 
  dplyr::left_join(data.frame(scaffold = names(scf_Olsson), width_scf = width(scf_Olsson))) %>% 
  group_by(transcript) %>% 
  mutate(n_match = n()) %>% 
  ungroup()
  
g1 <- ggplot(data_Olsson, aes(width_scf, width_trsc, col = n_match)) +
  geom_point() +
  xlab("Width of scaffolds from Olsson et al. (2016) (in bp)") +
  ylab("Width of transcripts from Tysklind et al. (in prep) (in bp)") +
  scale_color_continuous("Matches")
g2 <- ggplot(data_Olsson, aes(n_match)) + 
  geom_histogram(binwidth = 1) +
  xlab("Matches") +
  coord_flip()
cowplot::plot_grid(g1, g2, nrow = 1, rel_widths = c(3,1))
```
